src/Cli/dotnet/Commands/Run/FileBasedAppSourceEditor.cs

```diff
@@ -0,0 +1,223 @@
+        }
+    }
+
+    public required string NewLine { get; init; }
+
+    private FileBasedAppSourceEditor() { }
+
+    public static FileBasedAppSourceEditor Load(SourceFile sourceFile)
+    {
+        return new FileBasedAppSourceEditor
+        {
+            SourceFile = sourceFile,
+            Directives = LoadDirectives(sourceFile),
+            NewLine = GetNewLine(sourceFile.Text),
+        };
+
+        static string GetNewLine(SourceText text)
+        {
+            // Try to detect existing line endings.
+            string firstLine = text.Lines is [{ } line, ..]
+                ? text.ToString(line.SpanIncludingLineBreak)
+                : string.Empty;
+            return firstLine switch
+            {
+                [.., '\r', '\n'] => "\r\n",
+                [.., '\n'] => "\n",
+                [.., '\r'] => "\r",
+                [.., '\u0085'] => "\u0085",
+                [.., '\u2028'] => "\u2028",
+                [.., '\u2029'] => "\u2029",
+                _ => Environment.NewLine,
+            };
+        }
+    }
+
+    private static ImmutableArray<CSharpDirective> LoadDirectives(SourceFile sourceFile)
+    {
+        return VirtualProjectBuildingCommand.FindDirectives(sourceFile, reportAllErrors: false, DiagnosticBag.Ignore());
+    }
+
+    public void Add(CSharpDirective directive)
+    {
+        TextSpan span = DetermineWhereToAdd(directive, out var append);
+        string newText = append.Prefix + directive.ToString() + NewLine + append.Suffix;
+        SourceFile = SourceFile.WithText(SourceFile.Text.Replace(span, newText: newText));
+    }
+
+    private TextSpan DetermineWhereToAdd(CSharpDirective directive, out (string? Prefix, string? Suffix) append)
+    {
+        append = default;
+
+        // Find one that has the same kind and name.
+        // If found, we will replace it with the new directive.
+        if (directive is CSharpDirective.Named named &&
+            Directives.OfType<CSharpDirective.Named>().FirstOrDefault(d => NamedDirectiveComparer.Instance.Equals(d, named)) is { } toReplace)
+        {
+            return toReplace.Info.Span;
+        }
+
+        // Find the last directive of the first group of directives of the same kind.
+        // If found, we will insert the new directive after it.
+        CSharpDirective? addAfer = null;
+        foreach (var existingDirective in Directives)
+        {
+            if (existingDirective.GetType() == directive.GetType())
+            {
+                addAfer = existingDirective;
+            }
+            else if (addAfer != null)
+            {
+                break;
+            }
+        }
+
+        if (addAfer != null)
+        {
+            return new TextSpan(start: addAfer.Info.Span.End, length: 0);
+        }
+
+        // Otherwise, we will add the directive to the top of the file.
+        int start = 0;
+
+        var tokenizer = VirtualProjectBuildingCommand.CreateTokenizer(SourceFile.Text);
```

Comment:
IMO, this is an infrequent enough occurrence, that it would be fine to just parse, if it would make this implementation simpler. However, there's no strong need to make a change to do that.